### ISSUE TITLE
VMPC sampler corrupts 44.1k session input (Linux only?)

### ISSUE THREAD

**Issue by brownounces** [izzyreal/mpc#210]
When using Ubuntu Studio (but also observed on Gentoo; possibly other distros as well), it appears VMPC's SAMPLE function does not respect 44.1k sessions -- at least when set by JACK -- and corrupts the input after capturing the sample. We can isolate this issue to being a problem with the SAMPLER because non-sampled playback is unaffected (e.g. sounds loaded from the file browser, or dragged onto a pad, even when the sample is 16-bit/44.1k)

**First, let's copy my configuration from a fresh Ubuntu Studio install:**

1. Copy the following lines into a  terminal to disable PipeWire and return to ALSA/PulseAudio -- I don't consider PipeWire production-viable because the latency is considerably worse than JACK, causing artifacts while sampling and such, so let's axe it from the system. Some of these lines may or may not whine about permissions, those lines may need to be prefixed with 'sudo'

```
systemctl --user unmask pulseaudio
systemctl --user --now disable pipewire-media-session.service
systemctl --user --now disable pipewire pipewire-pulse
systemctl --user --now enable pulseaudio.service pulseaudio.socket
apt remove pipewire-audio-client-libraries pipewire
```

2. I would restart at this point, so we can activate PulseAudio and take control of our soundcard again.

3.  If possible, set your sound card to Analog Stereo Duplex -- this is the generic setting to render audio in 44.1k.

4. in ~/home/.config/pulse -- create a file called 'daemon.conf' with the following code block. I'm not sure how much this config file actually matters, but I use it to keep definition of what I am *expecting* of the PulseAudio sound server and my soundcard.

```
default-sample-format = s16le
default-sample-rate = 44100
realtime-scheduling = yes
exit-idle-time = -1
```

3. Now, we need the PulseAudio JACK module to create a sink to sample from. Type the following in your terminal:

`apt-get install pulseaudio-module-jack`

6. Open up qjackctl and define the following settings

```
Driver: alsa 

Interface: (your primary interface)
Sample Rate: 44100
Frames/Period: 1024
Periods/Buffer: (default)
```

>>  Advanced

`[x] Force 16bit`

>> Options

```
[x] Execute script after Startup

pactl load-module module-jack-sink channels=2; pactl load-module module-jack-source channels=2; pacmd set-default-sink jack_out
```

5. Now your JACK session is ready to go, and upon pressing Start, you enter a 16-bit/44.1k session and a PulseAudio sink should be spawned, allowing you to sample all the sound playing on the PulseAudio sound server, much like WASAPI Loopback or Stereo Mix does for Windows.

**--- Steps to Recreate on VMPC now that your setup is identical to mine ---**

1. Open up VMPC2000XL and select "JACK" as your audio device.

2. Set "system" as your Output and "PulseAudio JACK Sink" as your Input.

3. Attempt to sample anything with the SAMPLE function on VMPC; all should appear normal -- the monitor levels should all correspond correctly to the sound being played and all will look normal,

4. Attempt to playback the sample; it will have some non-descript, intermittent, or artifacted audio with no resemblence to what you sampled.

5. Examining the sample under the TRIM function reveals that the sample is as it sounds: some sparse artifacted audio that seems to be the result of some kind of sample rate dissonance.

6. Close VMPC2000XL, open up qjackctl, and define the sample rate as 48000, and start the session.

7. Attempt to sample a sound with the SAMPLE function; it will now sound 1:1 as recorded, seemingly pointing to a sample rate discrepancy between the SAMPLE function and VMPCs playback.

**--- Other notes ---**
The only thing affected seems to be sounds acquired through the SAMPLE function, _after_ the sound is saved into memory. All other sounds played back in a 44.1k JACK session are true to their original audio.

Bit depth seems to be respected by the sampler, or simply unaffected: running JACK at 32-bit or 16-bit yields identical results on both sample rate settings.

Why this is important: low-end systems may struggle to run 48k sessions, especially when DAWs get involved. With VMPC alone this is not normally a problem, but when used as a plugin in a DAW that is rendering every plugin and instrument in 48k, a basic production session can begin being taxing to a low end machine and frequent crashing can result.

Also: the original MPC is a 16-bit, 44.1khz sampler. To keep parity with the original machine, all features should be able to be rendered in 16/44.1 (if this is not some isolated bug with JACK or something) 😛

**Comment by izzyreal** [comment]
That looks great @brownounces, highly appreciated. Will get my hands dirty in the near future. Which version of Ubuntu Studio do you use for this?

**Comment by izzyreal** [comment]
At step 5, pressing Start in Qjackctl, I get https://pastebin.pl/view/13cf89eb

**Comment by izzyreal** [comment]
Hm, wait, after a restart it changed. I'll update here soon.

**Comment by izzyreal** [comment]
Yeah no still errors. I can't get Jack to start. Have you tried your steps in a fresh Ubuntu Studio VM? I've tried it with Ubuntu Studio 22 in VirtualBox. I'll play a bit with whatever comes to mind, but it would be great to get the steps to reproduce confirmed.

**Comment by izzyreal** [comment]
Ah it starts after using different settings for the VM:
* Host Audio Driver: ALSA Audio
* Audio Controller: Intel HD Audio
* Enable Audio Output
* Enable Audio Input

**Comment by izzyreal** [comment]
Yep, I can reproduce it :) Nicely. I'll do a debug build and see what's up.

**Comment by brownounces** [comment]
Sorry I couldn't respond earlier, only just got home -- glad you were able to get it working and reproduce it though!

**Comment by izzyreal** [comment]
I think I found the culprit and addressed it on this branch: https://github.com/izzyreal/mpc/tree/fix-sampling

Are you still using home-built VMPC2000XL? If yes, can you test with that branch?

The other thing is I'm getting weird read-outs in the waveform view in the TRIM screen:
![image](https://github.com/izzyreal/mpc/assets/3707432/bde4ef5b-6148-4850-bbca-48186d360b30)

The sounds sound fine, so it seems to be just cosmetic. Do you have similar visual artifacts?

**Comment by brownounces** [comment]
> Are you still using home-built VMPC2000XL? If yes, can you test with that branch?

Right now, I'm using the latest binary provided on izmar.nl, but I'll build a source copy to test and give you an update!


> The other thing is I'm getting weird read-outs in the waveform view in the TRIM screen:

I noticed this too, even when I was on Gentoo a while back. I'm not sure _why_ or under which conditions it appears this way  -- it still looks normal on your Arch and Ubuntu installs, right? 

It is perhaps an issue with the way the waveform is being rendered when two sound servers are running concurrently -- this sort of makes sense, because it appears the waveform render is demonstrating some kind of phase-cancelling, which you would expect when there is *technically* two servers providing the same audio. In this setup though, PulseAudio is just piping audio to JACK, and the render does not actively consider this (and why should it -- this is an unorthodox setup for people that belong in an insane asylum for the most part 😛).

**Comment by brownounces** [comment]
No dice on the fix on a home compiled VMPC, the result is the same. Tried both a 16-bit and a 32-bit JACK session. Still only likes 48000!

**Comment by izzyreal** [comment]
@brownounces Did you use https://github.com/izzyreal/mpc/tree/fix-sampling ?

**Comment by izzyreal** [comment]
@brownounces I pushed the fix to `master` branch as well now. Quite confident it should fix it. Once I saw the bug in my code I was "Ok, this is really bad" and it can actually be reproduced on any 44.1KHz setup 😅

**Comment by brownounces** [comment]
> @brownounces Did you use https://github.com/izzyreal/mpc/tree/fix-sampling ?

I git pulled the editables/mpc tree, but it didn't seem to have any effect

A fresh build though, with your latest commit, seems to have indeed fixed this!

>Once I saw the bug in my code I was "Ok, this is really bad" and it can actually be reproduced on any 44.1KHz setup

In your defense, most people are probably not running these very specific JACK settings or forcing 44.1k audio for performance gains, lol. I assume most people using VMPC are on Mac or Windows where low-latency audio is a long-solved issue (no need to mess around with forcing sample rates to squeeze every bit of performance out of an Ardour session), and the other three people using Linux are running PipeWire with brolic enough hardware to strong-arm it into being production viable.

In any case, your dedication to covering all the potential bases is A++. Being able to strictly enforce a 44.1khz session under any circumstances just makes it that much closer to a 1:1 2000XL emulator.

**Comment by izzyreal** [comment]
Resolved by f3a0bd68138629659feed045f1c562c849fb62ec eb45b3567ee7dd7c7e7b42f8bf685870d87cb978 39c9db67cd552a91db9ae39f3af4480f2485f83e 7aad7ab0ae38447490c1e42544bae1ae9251bcd2
